assembly/src/parsers/mod.rs

```diff
@@ -257,9 +257,13 @@ pub fn parse_module(source: &str) -> Result<ModuleAst, ParsingError> {
     };
     context.parse_procedures(&mut tokens, true)?;
 
-    // make sure program body is absent and no more instructions.
-    if tokens.read().is_some() {
-        return Err(ParsingError::unexpected_eof(tokens.pos()));
+    // make sure program body is absent and there are no more instructions.
+    if let Some(token) = tokens.read() {
+        if token.parts()[0] == Token::BEGIN {
```

Comment:
Maybe we need to check more variants since any of the following aren't allowed before `end`. The keywords that will fall into this example are:

`USE | EXPORT | PROC | BEGIN`

This means we likely should have a function `is_body_delimiter`